Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+The project is all of these files collected and processed by the implementing program together.
+
+### Comments
+
+A comment is a line that starts with the `\\` symbol.
+All text from the start of the comment to the end of the line must be ignored.
+A comment starting in the middle of another line ends that line at the point the `\\` symbol is encountered.
+That line is assumed to have finished at that point as if the comment was not there.
+Comments must not impact the rest of the lines or have any impact on the resulting yarn program.
+ 
+## Yarn Structure
+
+A Yarn script file is yarn file that contains one or more nodes and zero or more file tags.
+
+### File Tags
+
+_File tags_ are file level metadata that is relevant for all nodes in the file.
+A common use for file tags is for versioning the file.
+File tags must go at the start of a file before any nodes begin.
+File tags must have the `#` symbol at the start of them but are syntactically identical otherwise to the Header Tags of nodes.
+
+### Nodes
+
+A *node* is the single story element of a yarn file.
+Nodes are the story structural building blocks for yarn.
+Nodes are designed to contain pieces of a story and then have these story pieces linked together.
+This is not a requirement, everything could be done in a single node, this would just be unwieldy.
+A node must be comprised of a single header and a single body in that order.
+
+### Headers
+
+A header is comprised of one or more header tags.
+The header is finished when encountering a line that only contains the *header delimiter* `---`.
+After encountering the header delimiter the body of the node is entered.
+
+#### Header Tags
+
+A *header tag* is a line broken up into three components, in order; the tag name, the separator, and the tag text.
+The *tag name* can be any combination of the characters `A` - `Z`, `a` - `z`, `0` - `9`.
+The *tag separator* is the character `:`.
+The *tag text* can be any combination of the characters `A` - `Z`, `a` - `z`, `0` - `9`.
+Header tags are commonly used as node specific metadata but using them in this manner is not required.
+
+The amount of allowed whitespace between the tag name, the seperator, and the tag text is unspecified.
+An example of a header tag is the title tag: `Title:start`.
+
+Every node must have a title tag.
+Required or banned header tags beyond title are unspecified.
+The order of header tags is unspecified.
+
+#### Title Tag
+
+The *title tag* is a specific header tag that uniquely identifies the node.
+The tag name for the title tag must be `Title`.
+The tag text for the title tag must be unique within the file.
+The tag text for the title tag should be unique within the project.
+
+The behaviour of the program when a title tag's text is not unique across the project is unspecified.
+The program could either flag this as an error and abort, choose which node will be the canonical one with that title, or some other form of name conflict resolution.
```

Comment:
Iffy to leave this unspecified. A discussion on what exactly to choose as a solution is perhaps best left to another venue than comments on this file, but if Yarn requires nodes to be globally unique across the projec then, for the time being, I would encourage an exception to be thrown when the project is loaded. Anything else may be deceptive or confusing for the end-user, and if future specifications are stricter on this matter then forwards compatibility will become a big problem for users updating from Yarn v2

Reply:
The current implementation in Yarn Spinner for Unity, as of 2.0 beta 1, throws an error when two nodes in the same program have the same title, though a project can have multiple independent programs. (This isn't to say I think that this should be the Correct Behaviour or even that it should inform this spec, just that's what it's currently doing.)

Reply:
the reason I left this one like this is because I have seen people want to use the same node name in multiple files, so like `Title: Intro` in multiple characters and the resolution offered was to automatically append `filename-` to each title becoming `Title: filename-intro` and that actually feels like a reasonable thing you might want the system to handle for you.

As such I wanted to leave that possibility open, worth probably changing the recommendation to be an error by default though.